File: tests/csharp/tests-csharp-nunit-allure3-selenium/Tests/Infra/LocalBrowserPinTest.cs
using Allure.Net.Commons;
using Allure.NUnit.Attributes;
using Config;
using Helpers;
using Tests;

namespace Tests.Infra;

[AllureLabel("layer", "infra")]
[AllureEpic("Test infra")]
[AllureFeature("Local browser pin")]
[AllureSeverity(SeverityLevel.normal)]
[Category("infra")]
[Category("infra_frontend")]
[AllureSuite("Local browser pin")]
[NonParallelizable]
public sealed class LocalBrowserPinTest : AllureMeta
{
    private static string Major(string version) => version.Split('.')[0];

    [Test]
    [AllureName("pinnedVersion is a full Chrome for Testing build number")]
    public void PinnedVersionIsFullBuildNumber()
    {
        Assert.That(LocalChromePin.PinnedVersion(), Does.Match(@"^\d+\.\d+\.\d+\.\d+$"));
    }

    [Test]
    [AllureName("configured browserVersion stays on the pinned major")]
    public void ConfiguredBrowserVersionMatchesPin()
    {
        Assert.That(
            Major(ConfigReader.TestConfig.BrowserVersion),
            Is.EqualTo(Major(LocalChromePin.PinnedVersion())),
            "browserVersion and chrome-for-testing.properties drifted apart");
    }

    [Test]
    [AllureName("apply rejects a browserVersion from another major")]
    public void ApplyRejectsForeignMajor()
    {
        var foreignMajor = int.Parse(Major(LocalChromePin.PinnedVersion())) + 1;
        var error = Assert.Throws<InvalidOperationException>(() => LocalChromePin.Apply(foreignMajor.ToString()));
        Assert.That(error!.Message, Does.Contain("pinned build is"));
    }

    [Test]
    [AllureName("apply refuses to fall back to system Chrome")]
    public void ApplyRejectsBlankBrowserVersion()
    {
        var error = Assert.Throws<InvalidOperationException>(() => LocalChromePin.Apply(" "));
        Assert.That(error!.Message, Does.Contain("browserVersion is required"));
    }

    [Test]
    [AllureName("runtime rejects a non-Chrome browser")]
    public void RequireChromeRejectsFirefox()
    {
        var config = new TestConfig { Browser = "firefox" };
        var error = Assert.Throws<InvalidOperationException>(() => WebDrivers.RequireChrome(config));
        Assert.That(error!.Message, Does.Contain("Chrome-only"));
    }
}

File: tests/go/tests-go-testing-allure3-playwright/tests/infra/local_browser_pin_test.go
package infra_test

import (
	"regexp"
	"strconv"
	"testing"

	tests "tests-go-testing-allure3-playwright"

	allure "github.com/allure-framework/allure-go/commons/gotest"
	"github.com/stretchr/testify/require"
)

func wrapPin(t *testing.T, name string, body func(*allure.Context)) {
	t.Helper()
	tests.Wrap(t, name, body, tests.LayerInfraFrontend("Local browser pin", "Test infra", "Local browser pin", "normal")...)
}

func TestPinnedVersionIsFullChromeForTestingBuildNumber(t *testing.T) {
	wrapPin(t, "pinnedVersion is a full Chrome for Testing build number", func(a *allure.Context) {
		require.Regexp(t, regexp.MustCompile(`^\d+\.\d+\.\d+\.\d+$`), tests.PinnedChromeVersion())
	})
}

func TestConfiguredBrowserVersionStaysOnThePinnedMajor(t *testing.T) {
	wrapPin(t, "configured browserVersion stays on the pinned major", func(a *allure.Context) {
		pinMajor := splitMajor(tests.PinnedChromeVersion())
		cfgMajor := splitMajor(tests.LoadPlaywrightSettings().BrowserVersion)
		require.Equal(t, pinMajor, cfgMajor)
	})
}

func TestApplyRejectsForeignMajor(t *testing.T) {
	wrapPin(t, "apply rejects a browserVersion from another major", func(a *allure.Context) {
		major, err := strconv.Atoi(splitMajor(tests.PinnedChromeVersion()))
		require.NoError(t, err)
		err = tests.ApplyChromePin(strconv.Itoa(major + 1))
		require.Error(t, err)
		require.Contains(t, err.Error(), "pinned build is")
	})
}

func TestApplyRefusesToFallBackToSystemChrome(t *testing.T) {
	wrapPin(t, "apply refuses to fall back to system Chrome", func(a *allure.Context) {
		err := tests.ApplyChromePin(" ")
		require.Error(t, err)
		require.Contains(t, err.Error(), "browserVersion is required")
	})
}

func TestRuntimeRejectsANonChromiumBrowser(t *testing.T) {
	wrapPin(t, "runtime rejects a non-Chromium browser", func(a *allure.Context) {
		err := tests.RequireChromium("firefox")
		require.Error(t, err)
		require.Contains(t, err.Error(), "Chromium-only")
	})
}

func splitMajor(version string) string {
	for i, ch := range version {
		if ch == '.' {
			return version[:i]
		}
	}
	return version
}
